### PR TITLE
Add unit test for FreeRTOS SMP

### PR DESCRIPTION
**Pull request by chinglee-iot**
Description
-----------
* Add unit test for FreeRTOS SMP

Checklist:
----------
- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by kar-rahul-aws — DISMISSED** [review]

**Comment by chinglee-iot** [comment]
The line and function coverage rate is higher than single core. The branch coverage drops 2% due to new update to kernel.
This will be fixed with new PR.

**Single core coverage** 

Filename | -- | -- | Line Coverage | -- | Functions | -- | Branches | --
-- | -- | -- | -- | -- | -- | -- | -- | --
event_groups.c |   |   | 94.4 % | 186 / 197 | 100.0 % | 16 / 16 | 72.0 % | 72 / 100
list.c |   |   | 100.0 % | 41 / 41 | 100.0 % | 5 / 5 | 100.0 % | 6 / 6
queue.c |   |   | 93.1 % | 772 / 829 | 100.0 % | 49 / 49 | 97.0 % | 456 / 470
stream_buffer.c |   |   | 100.0 % | 329 / 329 | 100.0 % | 26 / 26 | 95.0 % | 190 / 200
tasks.c |   |   | 95.7 % | 870 / 909 | 96.1 % | 73 / 76 | 94.9 % | 444 / 468
timers.c |   |   | 100.0 % | 244 / 244 | 100.0 % | 30 / 30 | 96.2 % | 76 / 79

**SMP coverage **
<html><body>


Filename | -- | -- | Line Coverage | -- | Functions | -- | Branches | --
-- | -- | -- | -- | -- | -- | -- | -- | --
event_groups.c |   |   | 95.7 % | 176 / 184 | 100.0 % | 16 / 16 | 72.0 % | 72 / 100
list.c |   |   | 100.0 % | 40 / 40 | 100.0 % | 5 / 5 | 100.0 % | 6 / 6
queue.c |   |   | 93.1 % | 772 / 829 | 100.0 % | 49 / 49 | 97.0 % | 456 / 470
stream_buffer.c |   |   | 94.2 % | 310 / 329 | 100.0 % | 26 / 26 | 95.0 % | 190 / 200
tasks.c |   |   | 98.7 % | 1502 / 1522 | 99.0 % | 99 / 100 | 92.9 % | 1015 / 1092
timers.c |   |   | 100.0 % | 244 / 244 | 100.0 % | 30 / 30 | 96.2 % | 76 / 79

**Review by ActoryOu — APPROVED** [review]
It looks good to me. Could you help elaborate more about the difference between `multiple_priorities_no_timeslice_mock` and `multiple_priorities_no_timeslice`?

**Comment by chinglee-iot** [comment]
> It looks good to me. Could you help elaborate more about the difference between `multiple_priorities_no_timeslice_mock` and `multiple_priorities_no_timeslice`?

Thank you for the feedback. I added a [README.md](https://github.com/chinglee-iot/FreeRTOS/blob/smp-dev-add-smp-unit-test/FreeRTOS/Test/CMock/smp/README.md) file to describe the folder structure and naming convention.

**Review by ActoryOu — DISMISSED** [review]

**Review by ActoryOu — DISMISSED** [review]

**Review by ActoryOu — DISMISSED** [review]

File: FreeRTOS/Test/CMock/smp/README.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
+# FreeRTOS Kernel SMP Unit Test
+
+FreeRTOS Kernel SMP unit test verifies the SMP scheduler logic in tasks.c which is enclosed by `configNUMBER_OF_CORES > 1`. The common scheduler logic for single core and SMP is still verified in FreeRTOS/FreeRTOS/Test/CMock/tasks.
```
Comment (aggarg) [inline FreeRTOS/Test/CMock/smp/README.md:3]: Can we break the lines at 80 chars like in the `### Coverage tests` section?

**Review by aggarg — DISMISSED** [review]

File: FreeRTOS/Test/CMock/smp/README.md (hunk 1 above)
Comment (chinglee-iot) [reply FreeRTOS/Test/CMock/smp/README.md]: Update in new commit.

File: FreeRTOS/Test/CMock/config/FreeRTOSConfig.h [hunk 2]
```diff
@@ -134,8 +134,6 @@ void vConfigureTimerForRunTimeStats( void );    /* Prototype of function that in
         }                                             \
     } while( 0 )
 
-#define mtCOVERAGE_TEST_MARKER()    __asm volatile ( "NOP" )
```
[the remark this answers is not in the document]
Comment (chinglee-iot) [reply FreeRTOS/Test/CMock/config/FreeRTOSConfig.h]: Add it back in new commit.

**Review by aggarg — APPROVED** [review]

**Pull request merged**